Convert `Products by Attribute` to TS

albarin opened the pull request:
Converts the `Products by Attribute` block.js file to TS.

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

### Manual Testing

1. Create a page and add a `Products by Attribute` block.
2. Save the page and check it renders correctly the default config in the frontend.
3. Edit the page again and make some changes to the block (hiding some content, changing columns number, ordering, etc.).
4. Save the page and check it renders properly with the new config.

albarin committed:
Convert `Products by Attribute` block to TS

tomasztunik commented on assets/js/blocks/products-by-attribute/block.tsx:
```diff
@@ -167,10 +165,10 @@ class ProductsByAttributeBlock extends Component {
 
-	render() {
-		const { attributes, name, setAttributes } = this.props;
+	const render = () => {
```
Let's simplify it and just return JSX

tomasztunik commented on assets/js/blocks/products-by-attribute/block.tsx:
```diff
@@ -208,24 +206,46 @@ class ProductsByAttributeBlock extends Component {
-	debouncedSpeak: PropTypes.func.isRequired,
-};
+	debouncedSpeak: ( message: string ) => void;
+}
```
let's move interface definitions above implementation, we follow this layout in most of our files

tomasztunik commented on assets/js/blocks/products-by-attribute/block.tsx:
```diff
@@ -208,24 +206,46 @@ class ProductsByAttributeBlock extends Component {
 	 * A callback to update attributes
 	 */
-	setAttributes: PropTypes.func.isRequired,
+	setAttributes: ( attributes: Record< string, unknown > ) => void;
```
`setAttributes` could be typed as `( attributes: Partial< Attributes > ) => void;

tomasztunik commented on assets/js/blocks/products-by-attribute/block.tsx:
```diff
@@ -208,24 +206,46 @@ class ProductsByAttributeBlock extends Component {
 }
 
-ProductsByAttributeBlock.propTypes = {
+interface Props {
```
I'd love if we extracted this as a shared type as we've talked about and potentially pushed it upstream to Gutenberg.

We could type attributes with a generic type.

The `debouncedSpeak` comes from `withSpokenMessages` HOC so it's not shared.

I've tried to follow what is provided by default and did a dump of props for this block. A lot of these are shortcuts to actions on the blocks data store or flags based on the editor state ie. `isSelected`.  If we had these typed I believe we'd feel more comfortable using these :) 

I think for the minimum we should provide `name`, `clientId`, `attributes`, `setAttributes`, `isSelected` - though having other typed would be nice as well. We can split it as a separate PR maybe as it might require a bit of research on the types and which (if any) are already typed in Gutenberg or in DT

```
{ 
   attributes: {alt: '', contentAlign: 'center', dimRatio: 50, editMode: false, focalPoint: false, …}
   clientId: "1d574829-a8a2-49c9-a887-6f5735d278a6"
   context: {}
   insertBlocksAfter: ƒ ()
   isSelected: false
   isSelectionEnabled: true
   mergeBlocks: ƒ ()
   name: "woocommerce/featured-product"
   onRemove: () => removeBlock(clientId)
   onReplace: ƒ ()
   setAttributes: ƒ ()
   toggleSelection: ƒ ()
```



One thing regarding the Attributes type name to keep in mind is that it is distinct from BlockAttributes that we pass to a block when registering which holds attribute configuration. This one is provided by `@types/wordpress__blocks`

tomasztunik commented on assets/js/blocks/products-by-attribute/block.tsx:
```diff
@@ -26,9 +24,9 @@ import { getSetting } from '@woocommerce/settings';
-	getInspectorControls() {
-		const { setAttributes } = this.props;
+const ProductsByAttributeBlock = ( props: Props ): JSX.Element => {
+	const getInspectorControls = () => {
```
I'm thinking about the value of having these render methods and if it wouldn't be better to extract this one, `<ProductsByAttributeInspectorControls />` and `<ProductsByAttributeEditMode />` as separate components rather than methods I believe this will make the code easier to follow.

tomasztunik commented on assets/js/blocks/products-by-attribute/block.tsx:
```diff
@@ -208,24 +206,46 @@ class ProductsByAttributeBlock extends Component {
+};
+
+interface Attributes {
+	align?: string;
```
This can be replaced with `BlockAlignment` from `@types/wordpress__blocks`

albarin committed:
Small tweaks after review

albarin committed:
Extract InspectorControls and EditMode into components

tomasztunik reviewed (changes requested):
Thanks! This makes it much easier to follow the code.

I've noticed something that went by probably because of the overload that we've mixed concerns here (and in the Handpicked Products as well). 

When we register a block we have the block component and the edit component. All the controls for the editor should be mounted in the Edit component and the editMode should be controlled there. 

Here we are using PHP for rendering on the front-end I think but for the sake of consistency between blocks I believe we should move the editor controls into the `edit.tsx` file even if it meant block.tsx would only render the `<ServerSideRender />` and the `<Disabled>` wrapper would be moved to the edit as well.

What do you think Alba?

I'm also not entirely happy with how we store the editMode flag as an attribute and save it on the block and it's passed to the front-end as well. Pretty sure we could do other attributes inspection and store it as a local state of the editor block - not sure if we want to make it part of this change

albarin committed:
Refactor: move the edit related code to the edit.jsx file

albarin committed:
Remove the `editMode` attribute and control the edit mode inside the edit component

albarin commented:
> Here we are using PHP for rendering on the front-end I think but for the sake of consistency between blocks I believe we should move the editor controls into the `edit.tsx` file even if it meant block.tsx would only render the `<ServerSideRender />` and the `<Disabled>` wrapper would be moved to the edit as well.
> 
> What do you think Alba?

Makes a lot of sense, working on it 👌 

> I'm also not entirely happy with how we store the editMode flag as an attribute and save it on the block and it's passed to the front-end as well. Pretty sure we could do other attributes inspection and store it as a local state of the editor block - not sure if we want to make it part of this change

I've deleted the `editMode` flag and added state to the component. You are right, it doesn't make sense for it to be an attribute.

albarin committed:
Remove edit mode

albarin committed:
Change `isEditing` depending on attributes are selected or not

tomasztunik commented on assets/js/blocks/products-by-attribute/edit-mode.tsx:
```diff
@@ -0,0 +1,78 @@
+ */
+import { Attributes } from './types';
+
+export interface Props {
```
we can have this one extend Props and just add `isEditing/setIsEditing`

`export interface EditModeProps extends Props { ... }`

tomasztunik commented on assets/js/blocks/products-by-attribute/types.ts:
```diff
@@ -0,0 +1,40 @@
+export interface Attributes {
+	align?: BlockAlignment;
+	attributes: Array< string >;
+	attrOperator: string;
```
had a quick look in the php this accepts only `'all' | 'any'` strings

tomasztunik commented on assets/js/blocks/products-by-attribute/types.ts:
```diff
@@ -0,0 +1,40 @@
+		rating: boolean;
+		button: boolean;
+	};
+	orderby: string;
```
this one based on php can be `'date' | 'popularity' | 'price_asc' | 'price_desc' | 'rating' | 'title' | 'menu_order'`

albarin committed:
Improved interfaces/props after review

albarin committed:
Add enum allowed values to `attrOperator` and `orderby` on the `block.json`

albarin committed:
Move `withSpokenMessages` to wrap the whole Edit block
